src/main.rs

```diff
@@ -344,34 +350,33 @@ impl FormatBamRecords {
         // @CO	10x_bam_to_fastq:R1(RX:QX,TR:TQ,SEQ:QUAL)
         let re = Regex::new(r"@CO\t10x_bam_to_fastq:(\S+)\((\S+)\)").unwrap();
         let text = String::from_utf8(Vec::from(reader.header().as_bytes())).unwrap();
-        let mut spec = HashMap::new();
-
-        for l in text.lines() {
-            match re.captures(l) {
-                Some(c) => {
-                    let mut read_spec = Vec::new();
 
+        text.lines()
+            .into_iter()
+            .filter_map(|l| {
+                re.captures(l).map(|c| {
                     let read = c.get(1).unwrap().as_str().to_string();
                     let tag_list = c.get(2).unwrap().as_str();
-                    let spec_elems = tag_list.split(',');
-                    for el in spec_elems {
-                        if el == "SEQ:QUAL" {
-                            read_spec.push(SpecEntry::Read)
-                        } else {
-                            let mut parts = el.split(':');
-                            let rtag = parts.next().unwrap().to_string();
-                            let qtag = parts.next().unwrap().to_string();
-                            read_spec.push(SpecEntry::Tags(rtag, qtag));
-                        }
-                    }
 
-                    spec.insert(read, read_spec);
-                }
-                None => (),
-            }
-        }
+                    let spec_entries = tag_list
+                        .split(',')
+                        .into_iter()
+                        .map(|el| {
+                            if el == "SEQ:QUAL" {
+                                SpecEntry::Read
+                            } else {
+                                let mut parts = el.split(':');
+                                let rtag = parts.next().unwrap().to_string();
+                                let qtag = parts.next().unwrap().to_string();
```

Comment:
This isn't exactly the same thing, if there's more than one `:`, however I suspect the intent was
```suggestion
                                let (rtag, qtag) = el.split_once(':').unwrap();
```
which is considerably more efficient.  If there could be another `:` then that's easy enough to deal with:
```suggestion
                                let (rtag, qtag) = el.split_once(':').unwrap();
                                let qtag = qtag.split_once(':').map_or(qtag, |(s, _)| s);
```

Reply:
We are using `itertools` already, so [next_tuple](https://docs.rs/itertools/0.10.3/itertools/trait.Itertools.html#method.next_tuple) is available:
`let (rtag, qtag) = el.split(':').map(|m| m.to_string()).next_tuple().unwrap();`

Reply:
We could do that but then we should at least use `splitn(3, ':')` to avoid splitting more than we have to.  The advantage of `split_once` is because it returns exactly 2 values it doesn't need to allocate the container on the heap.

Reply:
Oh but I see how that works.  Yeah, I keep forgetting about how rust overloads by return value.  So as long as the split iterator is lazy then that's fine (I keep shifting my mindset between rust and go, and go doesn't really do lazy iterators, meaning `split` is always a heap allocation)